datafusion/physical-expr/src/aggregate/groups_accumulator/bool_op.rs

```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use arrow::array::AsArray;
+use arrow_array::{ArrayRef, BooleanArray};
+use arrow_buffer::BooleanBufferBuilder;
+use datafusion_common::Result;
+
+use crate::GroupsAccumulator;
+
+use super::accumulate::NullState;
+
+/// An accumulator that implements a single operation over a
+/// [`BooleanArray`] where the accumulated state is also boolean (such
+/// as [`BitAndAssign`])
+///
+/// F: The function to apply to two elements. The first argument is
+/// the existing value and should be updated with the second value
+/// (e.g. [`BitAndAssign`] style).
+///
+/// [`BitAndAssign`]: std::ops::BitAndAssign
+#[derive(Debug)]
+pub struct BooleanGroupsAccumulator<F>
+where
+    F: Fn(bool, bool) -> bool + Send + Sync,
+{
+    /// values per group
+    values: BooleanBufferBuilder,
+
+    /// Track nulls in the input / filters
+    null_state: NullState,
+
+    /// Function that computes the output
+    bool_fn: F,
+}
+
+impl<F> BooleanGroupsAccumulator<F>
+where
+    F: Fn(bool, bool) -> bool + Send + Sync,
+{
+    pub fn new(bitop_fn: F) -> Self {
+        Self {
+            values: BooleanBufferBuilder::new(0),
+            null_state: NullState::new(),
+            bool_fn: bitop_fn,
+        }
+    }
+}
+
+impl<F> GroupsAccumulator for BooleanGroupsAccumulator<F>
+where
+    F: Fn(bool, bool) -> bool + Send + Sync,
+{
+    fn update_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indices: &[usize],
+        opt_filter: Option<&BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 1, "single argument to update_batch");
+        let values = values.get(0).unwrap().as_boolean();
```

Comment:
`values[0]` can be used instead of `values.get(0).unwrap(`

Reply:
Done